# Titanic passengers analysis
The purpose of this project is to investigate the Titanic passengers data set and to understand which factors contributed to death or survival of passengers.


In [2]:
from matplotlib import pyplot as plt
import seaborn
import numpy as np
import pandas as pd
import titanicutils as tu
%matplotlib inline

## Data wrangling
Let's import CSV file with passengers data and get understanding of it. First let's take a look at the beginning of data set.

In [3]:
passengers = pd.read_csv('data/titanic-data.csv')
rows_number = len(passengers.index)
print(f'The size of data set is {rows_number} rows.')
passengers.head()

The size of data set is 891 rows.


<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PassengerId</th>
      <th>Survived</th>
      <th>Pclass</th>
      <th>Name</th>
      <th>Sex</th>
      <th>Age</th>
      <th>SibSp</th>
      <th>Parch</th>
      <th>Ticket</th>
      <th>Fare</th>
      <th>Cabin</th>
      <th>Embarked</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>0</td>
      <td>3</td>
      <td>Braund, Mr. Owen Harris</td>
      <td>male</td>
      <td>22.0</td>
      <td>1</td>
      <td>0</td>
      <td>A/5 21171</td>
      <td>7.2500</td>
      <td>NaN</td>
      <td>S</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>1</td>
      <td>1</td>
      <td>Cumings, Mrs. John Bradley (Florence Briggs Th...</td>
      <td>female</td>
      <td>38.0</td>
      <td>1</td>
      <td>0</td>
      <td>PC 17599</td>
      <td>71.2833</td>
      <td>C85</td>
      <td>C</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>1</td>
      <td>3</td>
      <td>Heikkinen, Miss. Laina</td>
      <td>female</td>
      <td>26.0</td>
      <td>0</td>
      <td>0</td>
      <td>STON/O2. 3101282</td>
      <td>7.9250</td>
      <td>NaN</td>
      <td>S</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>1</td>
      <td>1</td>
      <td>Futrelle, Mrs. Jacques Heath (Lily May Peel)</td>
      <td>female</td>
      <td>35.0</td>
      <td>1</td>
      <td>0</td>
      <td>113803</td>
      <td>53.1000</td>
      <td>C123</td>
      <td>S</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>0</td>
      <td>3</td>
      <td>Allen, Mr. William Henry</td>
      <td>male</td>
      <td>35.0</td>
      <td>0</td>
      <td>0</td>
      <td>373450</td>
      <td>8.0500</td>
      <td>NaN</td>
      <td>S</td>
    </tr>
  </tbody>
</table>
</div>

There were more than 2000 passengers onboard Titanic, but we have only 891 rows in our data set, meaning we're working with a sample and not with entire population of passengers.

### Analysing NaN values
It's clear that some of the columns contain _NaN_ in some rows. Let's try to understand how many of the columns need additional filtration. 

In [4]:
nans_by_column = tu.nans_count_by_column(passengers)
nans_by_column

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PassengerId</th>
      <th>Survived</th>
      <th>Pclass</th>
      <th>Name</th>
      <th>Sex</th>
      <th>Age</th>
      <th>SibSp</th>
      <th>Parch</th>
      <th>Ticket</th>
      <th>Fare</th>
      <th>Cabin</th>
      <th>Embarked</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>177</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>687</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div>

_Age_, _Cabin_ and _Embarked_ columns contain NaN values. Let's transform these values to percentage:

In [5]:
nans_percent_by_column = nans_by_column[['Age', 'Cabin', 'Embarked']]\
    .applymap(lambda nans_count: round(float(nans_count) / rows_number * 100, 2))
nans_percent_by_column

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Age</th>
      <th>Cabin</th>
      <th>Embarked</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>19.87</td>
      <td>77.1</td>
      <td>0.22</td>
    </tr>
  </tbody>
</table>
</div>

* _Age_ column is only 80% full. We should be careful with our conclusions based on passenger ages and treat the data set as a sample, because absense of so many data could potentialy skew the results.
* _Cabin_ column is almost empty - less than 23 percent of values in it contain data. Probably we won't be able to use it in our considerations. 
* _Embarked_ column is almost fully set. Later, if we decide to use it for analysis, we could filter the data set by removing all the rows with _Embarked_ value not set. Hopefuly this won't affect the global picture, because the value of rows with _Embarked_ value ommitted is negligent, but anyways, we should treat the dataset with filtered by _Embarked_ column as a sample.

## Analysis of data types
Let's take a look at the data types of our data set:

In [6]:
passengers.dtypes.to_frame().transpose()

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PassengerId</th>
      <th>Survived</th>
      <th>Pclass</th>
      <th>Name</th>
      <th>Sex</th>
      <th>Age</th>
      <th>SibSp</th>
      <th>Parch</th>
      <th>Ticket</th>
      <th>Fare</th>
      <th>Cabin</th>
      <th>Embarked</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>int64</td>
      <td>int64</td>
      <td>int64</td>
      <td>object</td>
      <td>object</td>
      <td>float64</td>
      <td>int64</td>
      <td>int64</td>
      <td>object</td>
      <td>float64</td>
      <td>object</td>
      <td>object</td>
    </tr>
  </tbody>
</table>
</div>

It looks like all the columns that should contain numerical data were parsed correctly. 
Now let's take a closer look at _Sex_ column. We expect it to contain two types of values _female_ and _male_. Let's check this column:

In [7]:
passengers['Sex'].unique()

array(['male', 'female'], dtype=object)

Indeed, the values in _Sex_ column are of correct type.

## Analysis of survival based on ticket class

Let's try to understand if ticket class has anything to do with survival rate.


In [66]:
survived_by_class = passengers[['Survived', 'Pclass']].groupby('Pclass')
survived_by_class.mean().applymap(lambda x: round(x, 2))

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Survived</th>
    </tr>
    <tr>
      <th>Pclass</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>0.63</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.47</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.24</td>
    </tr>
  </tbody>
</table>
</div>

The ratios are certainly different with higher survival rates for higher ticket class. Let's try to understand these results are statistically significant. Let's first try to do it for first and third classes - if we get the result that the difference survival rates between the best and worst case is not statistically significant, it should be the same for pairs of first and second classes and second and third classes.

We'll be conducting one-tailed t-test because on independent sample:
* We care about the direction. We want to have an answer to the question if being a passenger of a first class you have higher chances of survival.
* t-test because we have only sample of all the passengers aboard titanic, and we don't know the population parameters.
* The sample is independent because we assume that passenger did not change the ticket class.

As independent variable we're taking the ticket class and survival as a dependent variable.

$$
H_0: \bar{\mu}_{firstClass} - \bar{\mu}_{thirdClass} \leq 0
$$
$$
H_A: \bar{\mu}_{firstClass} - \bar{\mu}_{thirdClass} > 0
$$


In [64]:
std_by_class = survived_by_class.std()
means_by_class = survived_by_class.mean()
means_by_class

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Survived</th>
    </tr>
    <tr>
      <th>Pclass</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>0.629630</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.472826</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.242363</td>
    </tr>
  </tbody>
</table>
</div>